addon/-private/system/store.js

```diff
@@ -2107,13 +2107,40 @@ Store = Service.extend({
         throw new EmberError(`No model was found for '${modelName}'`);
       }
 
-      // interopt with the future
-      let klass = getOwner(this).factoryFor ? factory.class : factory;
+      // interop with the future
+      const hasFactoryFor = getOwner(this).factoryFor;
+      let klass = hasFactoryFor ? factory.class : factory;
 
       assert(`'${inspect(klass)}' does not appear to be an ember-data model`, klass.isModel);
 
-      // TODO: deprecate this
-      klass.modelName = klass.modelName || modelName;
+      // TODO: deprecate klass.modelName
+      /*
+        In test environments it is common to re-use the same base model in multiple
+        tests and register it each time. Since that seems convenient and good
+        to continue allowing this ignores the case where modelName is set but is
+         the same.
+       */
+      if (klass.modelName && klass.modelName !== modelName) {
```

Comment:
In theory (for the test case), we could detect if the original container has been destroyed or changed and if so, just seamlessly transition without a warning?